# Introduction

In this guided project, we work with a dataset of used cars from *eBay Kleinanzeigen*, a classifieds section of the German eBay website. While the full dataset consisting of over 370,000 listings was cleaned an uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data), this guided project will use a subset of 50,000 observations that has also been dirtied by the DataQuest team in order to more closely mimic what could be expected from the originally scraped data set.

The set contains the following columns.
- `dateCrawled`: When the ad was first crawled. All other field values for the corresponding row were scraped on this date.
- `name`: Name of the car listing.
- `seller`: Whether the seller is a private owner or a dealer.
- `offerType`: The type of listing.
- `price`: The price on the ad to sell the car.
- `abtest`: Whether the listing is included in an A/B test.
- `vehicleType`: The vehicle type.
- `yearOfRegistration`: The year in which the car was first registered.
- `gearbox`: The transmission type.
- `powerPS`: The power of the car in [PS](https://www.carwow.co.uk/guides/glossary/what-is-horsepower).
- `model`: The car model name.
- `odometer`: The odometer reading on the car, in kilometers.
- `monthOfRegistration`: The month in which the car was first registered.
- `fuelType`: What type of fuel the car uses.
- `brand`: The brand of the car.
- `notRepairedDamage`: Whether or not the car has damage which is not yet repaired.
- `dateCreated`: The date on which the eBay listing was created.
- `nrOfPictures`: The number of pictures in the listing.
- `postalCode`: The postal code for the location of the vehicle.
- `lastSeenOnline`: When the crawler last saw this listing online.

The goal of this project is to clean the data and then use pandas to perform some basic initial analysis of the listings. To start, we first import the NumPy and pandas libraries, and then will attempt read the CSV file which contains the data and load it into a pandas DataFrame.

In [1]:
import numpy as np
import pandas as pd

In [2]:
auto_filepath = "autos.csv"

In [3]:
autos = pd.read_csv(auto_filepath)

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xdc in position 23: invalid continuation byte

When trying to import the file using the default encoding of UTF-8, we get a `UnicodeDecodeError`. This means that we should attempt to use the next two most popular encodings, Latin-1 and Windows-1252, to see if that resolves the error.

In [4]:
# Attempt to use Latin-1 encoding to resolve UnicodeDecodeError
autos = pd.read_csv(auto_filepath, encoding = "Latin-1")

It seems as if using Latin-1 encoding allowed us to successfully read and load the dataset into a pandas DataFrame. Next we explore the first few rows of the data as well as the info provided by the `DataFrame.info()` method.

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Looking at the first few rows of the data, one thing that stands out immediately is the fact that many of the qualitative variables, such as `vehicleType`, `gearbox`, `fuelType`, and `notRepairedDamage`, use German vocabulary. This makes sense given the fact that the dataset was obtained by scraping German eBay listings. To better interpret the meanings of the values in the columns which use German, we should translate them into English. In addition, there are columns, such as `price` and `odometer`, which represent numeric quantities but are currently stored as strings. These columns will need to be converted and then renamed to indicate the units which are being used. Also, the `dateCrawled` and `lastSeen` columns appear to currently be stored as strings as well. We may wish to convert them to pandas `Timestamp` objects to make use of the robust [set of tools available in pandas to work with time series and dates](https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html).

In [6]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

Looking at the information provided by `DataFrame.info()`, we see that a number of columns contain null entries. Those columns are `vehicleType`, `gearbox`, `model`, `fuelType`, and `notRepairedDamage`. For each column which contains null entries, we will need to decide between dropping the corresponding row, leaving the null entry as-is, or replacing it with a descriptive value.

# Cleaning Column Names

While the column names are perfectly workable in their current state, there are a few things that we can clean up in order to make them easier to work with.

1. Change the following column names to be less clunky.
    - `yearOfRegistration` to `registration_year`
    - `monthOfRegistration` to `registration_month`
    - `notRepairedDamage` to `unrepaired_damage`
    - `dateCreated` to `ad_created`
2. Change the rest of the column names from camelcase to snakecase, which is the prefered style for Python.

In [7]:
auto_columns = autos.columns.to_series()
columns_dict = {"yearOfRegistration": "registration_year", "monthOfRegistration": "registration_month",
               "notRepairedDamage": "unrepaired_damage", "dateCreated": "ad_created"}
# First replace clunky column names
auto_columns.replace(columns_dict, inplace = True)
# Use a regular expression to insert underscores in between camelcase letters
auto_columns = auto_columns.str.replace(r"([a-z])([A-Z])", lambda m: "_".join(m.group(1, 2)))
# Make all column names lowercase
auto_columns = auto_columns.str.lower()
auto_columns

dateCrawled                  date_crawled
name                                 name
seller                             seller
offerType                      offer_type
price                               price
abtest                             abtest
vehicleType                  vehicle_type
yearOfRegistration      registration_year
gearbox                           gearbox
powerPS                          power_ps
model                               model
odometer                         odometer
monthOfRegistration    registration_month
fuelType                        fuel_type
brand                               brand
notRepairedDamage       unrepaired_damage
dateCreated                    ad_created
nrOfPictures               nr_of_pictures
postalCode                    postal_code
lastSeen                        last_seen
dtype: object

Now that we have changed the column names to make them easier to work with and also have them follow Python's preferred style conventions we replace the original column names with the modified ones and re-inspect the first few rows of the `autos` dataframe.

In [8]:
autos.columns = auto_columns
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

# Initial Exploration and Cleaning

Now that we have cleaned up the column names, it is time to continue our initial exploration of the data to determine what the next cleaning steps will be. First we will look at some basic descriptive statistics for our columns.

In [9]:
autos.describe(include = "all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-21 16:37:21  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [10]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [11]:
autos["offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

First, we notice that the `seller` and `offer_type` columns are almost exclusively a single value. That means they are good candidates to drop as they won't have useful information for analysis. It also appears that the column `nr_of_pictures` is all zeros, since the minimum and maximum values in the column are both zero, so I will investigate that to confirm.

In [12]:
autos["nr_of_pictures"].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

The column `nr_of_pictures` is indeed all zeros, so that is another good candidate to drop.

Continuing along, the minimum and maximum values in the `registration_year` column should also be addressed, as they are 1000 and 9999, respectively. Cars didn't exist in the year 1000, let alone eBay, and the year 9999 is far off into the future.

In [13]:
# Calculate the most recent date an observation was collected.
pd.to_datetime(autos["date_crawled"]).max()

Timestamp('2016-04-07 14:36:56')

In [14]:
# Count the number of cars with registration years prior to 1906 or after 2016
autos.loc[~autos["registration_year"].between(1906, 2016), "registration_year"].count()

1972

In [15]:
autos.loc[~autos["registration_year"].between(1906, 2016), "registration_year"].value_counts()

2017    1453
2018     492
9999       4
5000       4
2019       3
9000       2
1800       2
6200       1
4500       1
8888       1
4800       1
2800       1
1001       1
1000       1
1111       1
1500       1
9996       1
5911       1
4100       1
Name: registration_year, dtype: int64

Looking a little more closely we first note that that the most recent year the data was scraped was 2016, so there any listings which have a `registration_year` value of 2017 or later should be addressed. There are 1,972 entries which have a registration year earlier than 1906 ([the first year in which German license plates were issued with a lettering plan](https://en.wikipedia.org/wiki/Vehicle_registration_plates_of_Germany#History)) or after 2016. Since this is a only about 4% of the entries, dropping the corresponding rows wouldn't have a large negative impact on further analysis down the line. The majority of the entries with questionable registration year have a registration year of either 2017 or 2018, so might make sense to simply impute those values (along with the three values of 2019), if we had a way of confirming that those wrong years were simply the results of typos, and then drop the rows with the more egregious errors.

In [16]:
# Count the number of cars with power rating less than 1ps or greater than 2000ps
autos.loc[~autos["power_ps"].between(1, 2000), "power_ps"].count()

5518

In [17]:
autos.loc[~autos["power_ps"].between(1, 2000), "power_ps"].value_counts()

0        5500
5867        1
8404        1
2729        1
17700       1
7511        1
16011       1
6226        1
3500        1
16312       1
6045        1
15001       1
2018        1
6512        1
15016       1
9011        1
3750        1
14009       1
4400        1
Name: power_ps, dtype: int64

Next, the `power_ps` column has suspicious minimum (0ps) and maximum (17,700ps) values as well. It's impossible for a car to have a power rating of 0ps (though it some of the earliest cars had [power ratings of less than 1 ps](https://jalopnik.com/ten-of-the-least-powerful-road-cars-ever-made-1674061551)), and even the most [powerful car](https://en.wikipedia.org/wiki/List_of_automotive_superlatives#Power) available as of January 2020 (the [Lotus Evija](https://en.wikipedia.org/wiki/Lotus_Evija)) has a power rating of 2,000ps. One possibility is that the cars with a power rating of 0ps were listed without a power rating, so the value of zero actually represents a missing value. About 11% of the entries have power ratings outside the range of 1ps to 2,000ps, with the vast majority (5,500 entries) having a power listed as 0ps. That's a decent chunk of the data, so while it is pretty safe to drop the rows with the other impossible power values, depending on what we want to do with the data we may wish to impute the rows with a power of 0ps using something like the mean, median, or mode power value of the data set.

In [18]:
autos["registration_month"].value_counts()

0     5075
3     5071
6     4368
5     4107
4     4102
7     3949
10    3651
12    3447
9     3389
11    3360
1     3282
8     3191
2     3008
Name: registration_month, dtype: int64

Moving on to the `registration_month` column, we see that has a suspicious minimum value of 0. There is no zero month, so that could either be due to typos for September (9) or October (10), or accidentally using zero indexing when counting months. It could also represent cars for which a registration month was not given or found by the crawler. About 10% of the rows have a `registration_month` value of 0, so once again while we could drop those rows, we should consider what kind of analysis we wish to do going forward in order to decide between dropping the rows in question, imputing their `registration_month` values, or leaving them as-is.

Lastly, while the minimum value of 1067 for the `postal_code` column initially looked potentially suspicious, it is reasonable given the [postal code system in Germany](https://en.wikipedia.org/wiki/List_of_postal_codes_in_Germany).

Before moving on to the `price` and `odometer` columns, which are numeric values stored as text, we once again recall that the `vehicle_type`, `gearbox`, `model`, `fuel_type`, and `unrepaired_damage` columns contain some null values. None of the columns contains more than 20% null values, but will will still need to clean those up if at some point if we want to do analysis involving those columns.

In [19]:
# First strip leading $
cleaned_price = autos["price"].str.strip("$")
# Remove commas
cleaned_price = cleaned_price.str.replace(",", "")
# Convert to float64 dtype
cleaned_price = cleaned_price.astype(float)
autos["price"] = cleaned_price

To clean up the `price` column, all we needed to do was strip the `$` character and then remove all commas before casting it to the `float64` dtype.

In [20]:
# First strip trailing "km"
cleaned_odometer = autos["odometer"].str.strip("km")
# Remove commas
cleaned_odometer = cleaned_odometer.str.replace(",", "")
# Convert to float64
cleaned_odometer = cleaned_odometer.astype(float)
autos["odometer"] = cleaned_odometer

Cleaning up the `odometer` column was quite similar to cleaning the `price` column, except instead of stripping the `$` character we needed to strip the trailing `km`. Now that we have cleaned up the `odometer` column, we rename it to `odometer_km` so we can keep the information that the odometer readings for each row were recorded in kilometers.

In [21]:
autos.rename(columns = {"odometer": "odometer_km"}, inplace = True)

# Exploring the Odometer and Price Columns

Now that we have computed the `odometer_km` and `price` columns from strings into floats, we can analyze them to look for potential outliers that we may want to either remove or impute, depending on the further analysis we wish to do. It is important, however, when analyzing these columns for potential outliers, to ask ourselves if any outliers we encounter might provide meaningful information before excluding those values or imputing them. Especially if our domain knowledge tells us that the outliers, while rare, are still possible, then they may be worth keeping as-is. First, we will explore the `odometer_km` column.

In [22]:
autos["odometer_km"].nunique()

13

In [23]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [24]:
autos["odometer_km"].value_counts()

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: odometer_km, dtype: int64

In [25]:
(autos["odometer_km"] >= 100000).sum()

39763

We see that there are 13 distinct values in the `odometer_km` column, ranging from 5,000km to 150,000km. This is a reasonable range of odometer readings for used cars, since it corresponds to a range of approximately 3,100-93,000 miles. The majority of cars (about 60%) in the set have an odometer reading of 150,000km, and almost 80% of the cars have an odometer reading of at least 100,000km. While I would expect more unique values in this column, I assume that the small number is due to rounding choices made by the person who originally collected this data or by limitations in the odometer reading options users have when posting a listing. Since there don't appear to be any outliers in the `odometer_km` column, we move on to the `price` column.

In [26]:
autos["price"].nunique()

2357

There are only 2,357 distinct values in the `price` column. Again this is a fairly small number relative to the size of the dataset, but I assume it is likely due to some of the more common preferences people have when setting prices (e.g. rounding to even multiples of 5, 10, 100, 500, 1000, etc.).

In [27]:
np.round(autos["price"].describe(), 2)

count       50000.00
mean         9840.04
std        481104.38
min             0.00
25%          1100.00
50%          2950.00
75%          7200.00
max      99999999.00
Name: price, dtype: float64

The `price` column has both unrealistic minimum and maximum values that immediately stand out. At the low end, it is impossible to post any item on eBay with a price of zero. At the high end, a price of \$99,999,999, while certainly a possible list price ([in 2005 a \$168 million gigayacht was listed and reportedly sold on eBay](https://www.ebayinc.com/company/our-history/)), it is still quite unrealistic. Looking through posts from eBay about the most expensive and interesting purchases made through the site ([2019](https://www.ebayinc.com/stories/news/ebay-unveils-the-most-interesting-and-expensive-purchases-of-2019/), [2018](https://www.ebayinc.com/stories/news/revealing-ebays-20-most-expensive-purchases-of-2018/), [2017](https://www.ebay.com/motors/blog/the-most-expensive-cars-sold-on-ebay-motors-in-2017/), [2016](https://www.ebay.com/motors/blog/most-expensive-cars-ebay-2016/), [2010](https://www.ebay.com/motors/blog/most-expensive-cars-sold-ebay/)), while an imperfect reference since it only discusses listings that were actually sold, suggests that the highest realistic price for a car would be \$4,000,000, and that typically the most expensive cars are listed for \$2,000,000 or less.

In [28]:
# Get value counts for the prices that are less than 0.01 and greater than 4000000
autos.loc[~autos["price"].between(0.01, 4000000), "price"].value_counts().sort_index()

0.0           1421
10000000.0       1
11111111.0       2
12345678.0       3
27322222.0       1
99999999.0       1
Name: price, dtype: int64

In [29]:
# Count the number of cars with prices less than 0.01 and greater than 4000000
autos.loc[~autos["price"].between(0.01, 4000000), "price"].count()

1429

We see that the vast majority of cars with a unrealistic prices have a price of zero, with 1,421 such vehicles, and in total only about 3% (1,429) of the listings have extreme outlier prices (prices of either zero or larger than 4,000,000). That is a very small fraction of the listings, so even if those prices are accurate, it still makes sense to exclude those listings if we want to do any analyses that involve the price. Let's exclude those outlier prices to see how that affects the descriptive statistics for the remaining values.

In [30]:
np.round(autos.loc[autos["price"].between(0.01, 4000000), "price"].describe(), 2)

count      48571.00
mean        6082.24
std        22786.98
min            1.00
25%         1200.00
50%         3000.00
75%         7490.00
max      3890000.00
Name: price, dtype: float64

After excluding the most egregious prices, there still appears to be some rather extreme values at both the high and the low ends.

In [31]:
autos.loc[autos["price"].between(0.01, 4000000), "price"].value_counts().sort_index().head(25)

1.0     156
2.0       3
3.0       1
5.0       2
8.0       1
9.0       1
10.0      7
11.0      2
12.0      3
13.0      2
14.0      1
15.0      2
17.0      3
18.0      1
20.0      4
25.0      5
29.0      1
30.0      7
35.0      1
40.0      6
45.0      4
47.0      1
49.0      4
50.0     49
55.0      2
Name: price, dtype: int64

At the low end, there are prices that are below \$100. While those are quite low prices for a car, they are still within reason if there are sellers who are either trying to get rid of a truly junky car or are trying to use a very low price to catch the attention of potential buyers browsing the eBay listings.

In [32]:
autos.loc[autos["price"].between(0.01, 4000000), "price"].value_counts().sort_index().tail(25)

151990.0     1
155000.0     1
163500.0     1
163991.0     1
169000.0     1
169999.0     1
175000.0     1
180000.0     1
190000.0     1
194000.0     1
197000.0     1
198000.0     1
220000.0     1
250000.0     1
259000.0     1
265000.0     1
295000.0     1
299000.0     1
345000.0     1
350000.0     1
999990.0     1
999999.0     2
1234566.0    1
1300000.0    1
3890000.0    1
Name: price, dtype: int64

There are a small handful of cars at the high end. The prices seem to increase fairly incrementally up to \$350,000, and then there is a large jump from there. Since the cars with prices higher than \$350,000 could still be legitimate luxury vehicles or otherwise valuable cars, and since there are so few of them, we could inspect those rows by hand to see if the vehicle name, model, and brand provide any information to help us decide whether or not to exclude them, or if we might want to impute those price values with something more reasonable.

In [33]:
autos.loc[autos["price"] > 350000, ["name", "model", "brand", "registration_year", "price"]]

name        model  \
514                     Ford_Focus_Turnier_1.6_16V_Style        focus   
2897    Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000       escort   
7814                                         Ferrari_F40          NaN   
11137  suche_maserati_3200_gt_Zustand_unwichtig_laufe...          NaN   
22947                         Bmw_530d_zum_ausschlachten          NaN   
24384                            Schlachte_Golf_3_gt_tdi          NaN   
27371                                         Fiat_Punto        punto   
37585                                Volkswagen_Jetta_GT        jetta   
39377                        Tausche_volvo_v40_gegen_van          v40   
39705                        Tausch_gegen_gleichwertiges     s_klasse   
42221                                  Leasinguebernahme           c4   
43049                                      2_VW_Busse_T3  transporter   
47598  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...       vectra   
47634                                        Ferrari_FXX          NaN   

                brand  registration_year       price  
514              ford               2009    999999.0  
2897             ford               1973  11111111.0  
7814   sonstige_autos               1992   1300000.0  
11137  sonstige_autos               1960  10000000.0  
22947             bmw               1999   1234566.0  
24384      volkswagen               1995  11111111.0  
27371            fiat               2017  12345678.0  
37585      volkswagen               1985    999990.0  
39377           volvo               2018  12345678.0  
39705   mercedes_benz               1999  99999999.0  
42221         citroen               2014  27322222.0  
43049      volkswagen               1981    999999.0  
47598            opel               2001  12345678.0  
47634  sonstige_autos               2006   3890000.0

While there are some cars that clearly shouldn't have such a high price (such as the Ford Focus in the first row, unless it was posted with a price of \$999,999 as a joke) and others have prices that are conceivable (such as the Ferrari FXX in the last row), other cars would require more research to determing whether or not they have reasonable prices in this data set. Doing this might not be worth the time and effort unless we need accurate price information to answer questions about cars at the top end of the pricing spectrum. If we are more interested in analyzing the vast majority of cars with prices of \$350,000 or less, then it would be a lot easier to simply drop these rows.

# Exploring the Date Columns

Now that we have explored the `odometer_km` and `price` columns and identified some possible further tasks for cleaning up the data, we continue our exploration by looking at the five columns which represent date values: `date_crawled`, `last_seen`, `ad_created`, `registration_month`, and `registration_year`. As we saw in our initial exploration, the `date_crawled`, `last_seen`, and `ad_created` columns are currently reprented as strings, while the `registration_month` and `registration_year` columns are stored as integers. We don't need to do any additional preprocessing to compute helpful summary statistics for the `registration_month` and `registration_year` columns, so we will first focus our attention to the date columns which are stored as strings.

In [34]:
autos[["date_crawled", "ad_created", "last_seen"]].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

Looking at the head of those columns, it appears the strings are formatted as timestamps, so we can make use of the built-in tools that pandas has for working with dates and times. While we will only be using very basic functionality, there are helpful detailed explanations for the the date/time functionality in pandas in the [official documentation](https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html) and in the [*Python Data Science Handbook* by Jake VanderPlas](https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html). We will start off by converting those columns to pandas `datetime` objects using the `pandas.to_datetime()` function. More info about all of the optional arguments for that function can be found [here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_datetime.html#pandas.to_datetime).

In [35]:
autos[["date_crawled", "ad_created", "last_seen"]] = autos[["date_crawled", "ad_created", "last_seen"]].apply(pd.to_datetime)

In [36]:
autos[["date_crawled", "ad_created", "last_seen"]].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 3 columns):
date_crawled    50000 non-null datetime64[ns]
ad_created      50000 non-null datetime64[ns]
last_seen       50000 non-null datetime64[ns]
dtypes: datetime64[ns](3)
memory usage: 1.1 MB


Now that we have converted those columns to `datetime` objects, we can quite conveniently calculate the distributions of the values in a large variety of ways. For each column, we will start off by investigating the distributions based on full date (day, month, year), but we can also explore the distributions based on [year, day of month, day of week, hour of day, and more](https://pandas.pydata.org/pandas-docs/stable/reference/series.html#datetime-properties). In order to convert each full timestamp into the format `YYYY-MM-DD`, we take the [floor of each timestamp based on using days as my frequency](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.dt.floor.html#pandas.Series.dt.floor).

In [37]:
autos["date_crawled"].dt.floor("D").value_counts(normalize = True).sort_index()

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

We can see that the data was scraped over the period of approximately one month, from March 5, 2016 to April 7, 2016, with the distribution being fairly uniform. Since the data was fairly uniformly scraped over the course of the month, analyzing the distribution based on day of week wouldn't be particular interesting, but we can still look at the distribution based on hour.

In [38]:
autos["date_crawled"].dt.hour.value_counts(normalize = True).sort_index()

0     0.02716
1     0.01542
2     0.00652
3     0.00502
4     0.00420
5     0.00310
6     0.00350
7     0.01442
8     0.01994
9     0.03556
10    0.04812
11    0.05540
12    0.05884
13    0.06466
14    0.06808
15    0.06626
16    0.06960
17    0.06988
18    0.07356
19    0.06792
20    0.06780
21    0.06148
22    0.05088
23    0.04268
Name: date_crawled, dtype: float64

It appears that most of the data was scraped pretty uniformly outside the hours 2:00am to 6:59am. Now we move on to the `ad_created` column.

In [39]:
autos["ad_created"].dt.floor("D").value_counts(normalize = True).sort_index().head()

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
Name: ad_created, dtype: float64

In [40]:
autos["ad_created"].dt.floor("D").value_counts(normalize = True).sort_index().tail()

2016-04-03    0.03892
2016-04-04    0.03688
2016-04-05    0.01184
2016-04-06    0.00326
2016-04-07    0.00128
Name: ad_created, dtype: float64

In [41]:
autos["ad_created"].dt.year.value_counts(normalize = True)

2016    0.99988
2015    0.00012
Name: ad_created, dtype: float64

In [42]:
autos.loc[autos["ad_created"].dt.year == 2016, "ad_created"].dt.month.value_counts(normalize = True).sort_index()

1    0.000240
2    0.001260
3    0.837420
4    0.161079
Name: ad_created, dtype: float64

The ads were all created between June 11, 2015 and April 7, 2016 (the last date on which data was scraped). The vast majority of listings are from March and April 2016. Since I did so with the `date_crawled` column, we will also look at the distribution of when the ads were created based on hour.

In [43]:
autos["ad_created"].dt.hour.value_counts(normalize = True).sort_index()

0    1.0
Name: ad_created, dtype: float64

In [44]:
autos["ad_created"].dt.time.value_counts(normalize = True).sort_index()

00:00:00    1.0
Name: ad_created, dtype: float64

All of listings have a creation time of 00:00:00 (exactly midnight), which doesn't make sense. I assume it may be due to either only the date of creation being accessible to the crawler that scraped the data or a choice made by the person who collected the data to exclude the finer-grained time information for this column. For the purposes of overall analysis it doesn't make much a difference, though not having fine-grained time information does prevent us from exploring the possibility of potentially interesting time-based patterns related to when ads are listed. Lastly, we look over the `last_seen` column.

In [45]:
autos["last_seen"].dt.floor("D").value_counts(normalize = True).sort_index()

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64

In [46]:
autos["last_seen"].dt.floor("D").value_counts(normalize = True).sort_index().tail().sum()

0.52618

Just like the `date_crawled` column, the `last_seen` column covers the period from March 5, 2016 through April 7, 2016 during which the data was scraped. Over half of the entries were last seen in the last five days of data scraping. While the date on which an entry was last seen can be used to determine if an entry was removed (due to the car being purchased, the listing expiring, or the seller taking the listing down for other reasons) by comparing it to the corresponding date in the `date_crawled` column for that row, we shouldn't make those assumptions for cars which were seen in the last few days of scraping (April 5, 6, 7). It is likely that if a car was last seen on those days, it is simply due to the fact that that is the end of the scraping window. This is especially likely if the crawler didn't check every single listing it had previously seen on every day of crawling. 

Finally, as noted above in the Initial Exploration and Cleaning section, there are some odd values in both the `registration_month` and `registration_year` columns.

In [47]:
autos["registration_month"].value_counts(normalize = True).sort_index()

0     0.10150
1     0.06564
2     0.06016
3     0.10142
4     0.08204
5     0.08214
6     0.08736
7     0.07898
8     0.06382
9     0.06778
10    0.07302
11    0.06720
12    0.06894
Name: registration_month, dtype: float64

First, about 10% of the listings have a registration month of `0`, which is an impossible month. This could either be due to typos in the original listings, or in this context a value of zero could indicate that the listing didn't include the registration month.

In [48]:
autos["registration_year"].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [49]:
# Calculate the most recent date an observation was collected.
autos["date_crawled"].max()

Timestamp('2016-04-07 14:36:56')

In [50]:
# Count the number of cars with registration years prior to 1906 or after 2016
autos.loc[~autos["registration_year"].between(1906, 2016), "registration_year"].count()

1972

In [51]:
autos.loc[~autos["registration_year"].between(1906, 2016), "registration_year"].value_counts()

2017    1453
2018     492
9999       4
5000       4
2019       3
9000       2
1800       2
6200       1
4500       1
8888       1
4800       1
2800       1
1001       1
1000       1
1111       1
1500       1
9996       1
5911       1
4100       1
Name: registration_year, dtype: int64

Second, 1,972 entries (about 4% of the entries) have impossible registration years (prior to 1906 or after 2016, the year in which the data was collected).

# Dealing with Incorrect Registration Year Data

In the previous section we noted that about 4% of the entries have impossible registration years. Since that is a small proportion of the data it is fairly safe to simply drop those rows, though imputing those values might also be an option if we wish to take the time to analyze them a little more closely. For now, we'll explore the distribution of the `registration_year` column after removing the impossible years.

In [52]:
autos.loc[autos["registration_year"].between(1906, 2016), "registration_year"].describe()

count    48028.00000
mean      2002.80351
std          7.31085
min       1910.00000
25%       1999.00000
50%       2003.00000
75%       2008.00000
max       2016.00000
Name: registration_year, dtype: float64

The remaining registration years range from 1910 to 2016, with at least least 75% of the registrations coming from 1999 or later.

In [53]:
autos.loc[autos["registration_year"].between(1906, 2016), "registration_year"].quantile(0.1)

1995.0

In [54]:
autos.loc[autos["registration_year"].between(1995, 2016), "registration_year"].value_counts(normalize = True).sort_index()

1995    0.029579
1996    0.032531
1997    0.045687
1998    0.055261
1999    0.067584
2000    0.075559
2001    0.060893
2002    0.057064
2003    0.061434
2004    0.061659
2005    0.067922
2006    0.061006
2007    0.051905
2008    0.050260
2009    0.047264
2010    0.035977
2011    0.036811
2012    0.029805
2013    0.018158
2014    0.015004
2015    0.008989
2016    0.029647
Name: registration_year, dtype: float64

Looking a little closer, 90% of the registrations come from 1995 or later, and those registrations are roughly evenly distributed.

In [55]:
autos.loc[autos["registration_year"].between(1906, 1994), "registration_year"].value_counts(normalize = True).sort_index()

1910    0.002473
1927    0.000275
1929    0.000275
1931    0.000275
1934    0.000550
1937    0.001099
1938    0.000275
1939    0.000275
1941    0.000550
1943    0.000275
1948    0.000275
1950    0.000824
1951    0.000550
1952    0.000275
1953    0.000275
1954    0.000550
1955    0.000550
1956    0.001374
1957    0.000550
1958    0.001099
1959    0.001924
1960    0.009343
1961    0.001649
1962    0.001099
1963    0.002473
1964    0.003298
1965    0.004672
1966    0.006046
1967    0.007420
1968    0.007145
1969    0.005221
1970    0.012366
1971    0.007420
1972    0.009618
1973    0.007145
1974    0.006595
1975    0.005221
1976    0.007420
1977    0.006046
1978    0.012916
1979    0.009618
1980    0.026656
1981    0.008519
1982    0.011816
1983    0.014564
1984    0.014564
1985    0.028854
1986    0.020885
1987    0.020610
1988    0.039022
1989    0.049739
1990    0.108546
1991    0.097829
1992    0.107447
1993    0.122286
1994    0.181369
Name: registration_year, dtype: float64

In [56]:
autos.loc[autos["registration_year"].between(1906, 1994), "registration_year"].value_counts(normalize = True).sort_index().loc[1970:].sum()

0.9370706237977465

The pre-1995 registrations are much less evenly distributed, with about 60% of those registrations still coming from the 1990s, and almost 94% of them coming from the 1970s or later.

In [57]:
autos.loc[autos["registration_year"].between(1906, 2016), "registration_year"].value_counts(normalize = True).sort_index().loc[1970:].sum()

0.9952319480303154

In fact, 99.5% of the cars have registration years from 1970 or later.

# Initial Exploration and Cleaning Summary

After reading the data from the `autos.csv` file and looking at the first few rows, we cleaned up the column names to make them less clunky and also convert them from camelcase to snakecase in order to follow Python's style recommendations. We then explored each column for null values, numeric data that needs conversion into numeric types, and outliers among the numeric columns. We also looked for columns that contained almost exclusively a single value, since those aren't likely to be helpful for any analysis we wish to do. The columns that needed particular attention were as follows.

- `seller`, `offer_type`, `nr_of_pictures`: Contain almost exclusively a single value. We should drop these columns.
- `vehicle_type`, `gearbox`, `model`, `fuel_type`, `unrepaired_damage`: Contain missing values, though none has more than 20% missing values. Depending on what we want to do with this data, we can drop the rows with missing values, impute the missing values, or leave them as-is. In addition, the `vehicle_type`, `gearbox`, `fuel_type`, and `unrepaired_damage` columns contain categorical data that uses German words, so those can be translated into English for easier comprehension.
- `date_crawled`, `ad_created`, `last_seen`: Convert from string to pandas `datetime` type.
- `price`: Convert from string to float. About 3% of values are impossible (a price of \$0) and 14 values are potentially unrealistic (prices exceeding \$350,000). It is safe to drop the rows with impossible prices, but we should think more carefully before dropping or imputing the unrealistic ones depending on the analysis we wish to do.
- `registration_year`: About 4% of the values are impossible (prior to 1906 or after 2016). It is fairly safe to drop the rows with these impossible registration years.
- `registration_month`: About 10% of the values are impossible (0). We should consider either dropping the corresponding rows, imputing the values, or leaving them as-is, depending on the analysis we wish to do.
- `power_ps`: About 11% of the entries are impossible (either 0 or exceeding 2000). We should consider dropping the rows with impossible power ratings or imputing the values depending on how we want to use this column.
- `odometer_km`: Convert from string to float and rename column to preserve units information.

For the time being, we will drop the `seller`, `offer_type`, and `nr_of_pictures` columns. Then we drop the rows with impossible `price` and `registration_year` values. We also drop the rows with the most unrealistic `price` values (exceeding \$4,000,000) at this point in time. Finally, we will translate the `vehicle_type`, `gearbox`, `fuel_type`, and `unrepaired_damage` columns from German to English.

In [58]:
# Drop seller, offer_type, nr_of_pictures columns
autos.drop(columns = ["seller", "offer_type", "nr_of_pictures"], inplace = True)

In [59]:
# Filter out rows with impossible registration year 
# and impossible/unrealistic price
price_year_filter = autos["registration_year"].between(1906, 2016) & autos["price"].between(0.01, 4000000)
autos = autos[price_year_filter]

After dropping the rows and columns we chose to drop, we turn our attention to translating the categorical columns from German into English. First up is the `vehicle_type` column.

In [60]:
autos["vehicle_type"].value_counts()

limousine     12599
kleinwagen    10585
kombi          8932
bus            4032
cabrio         3016
coupe          2464
suv            1965
andere          390
Name: vehicle_type, dtype: int64

There eight different vehicle types. While some, such as coupe or SUV don't need to be translated into English, the others require translation. Using a mix of Google Translate, the [English](https://en.wikipedia.org/wiki/Car_classification)/[German](https://de.wikipedia.org/wiki/Fahrzeugklasse) Wikipedia pages for car classification,the [English](https://en.wikipedia.org/wiki/Car_body_style)/[German](https://de.wikipedia.org/wiki/Karosseriebauform) Wikipedia pages for car body styles and [eBay Kleinanzeigen](https://www.ebay-kleinanzeigen.de/), we have the following translations:

- andere $\Leftrightarrow$ other
- bus $\Leftrightarrow$ van
- cabrio $\Leftrightarrow$ convertible
- kleinwagen $\Leftrightarrow$ supermini (European classification) or subcompact (American classification)
- kombi $\Leftrightarrow$ station wagon
- limousine $\Leftrightarrow$ sedan

In [61]:
vehicle_types = {"andere": "other", "bus": "van", "cabrio": "convertible",
                "coupe": "coupe", "kleinwagen": "subcompact", "kombi": "station wagon",
                "limousine": "sedan", "suv": "suv"}
autos["vehicle_type"] = autos["vehicle_type"].map(vehicle_types)

In [62]:
autos["gearbox"].value_counts()

manuell      34718
automatik     9857
Name: gearbox, dtype: int64

There are only two different gearbox types, which are very straightforward to translate.

In [63]:
autos["gearbox"] = autos["gearbox"].map({"manuell": "manual", "automatik": "automatic"})

In [64]:
autos["fuel_type"].value_counts()

benzin     28543
diesel     14033
lpg          649
cng           71
hybrid        37
elektro       19
andere        15
Name: fuel_type, dtype: int64

There are seven different fuel types. Again, using Google translate gives us the following translations. We also refer to the [Wikipedia page about alternative fuel vehicles](https://en.wikipedia.org/wiki/Alternative_fuel_vehicle) for explanations of CNG (compressed natural gas) and LPG (liquefied petroleum gas). There are only three types that we need to translate.

- andere $\Leftrightarrow$ other
- benzin $\Leftrightarrow$ gasoline
- elektro $\Leftrightarrow$ electric

In [65]:
fuel_types = {"andere": "other", "benzin": "gasoline", "cng": "cng", "diesel": "diesel",
             "elektro": "electric", "hybrid": "hybrid", "lpg": "lpg"}
autos["fuel_type"] = autos["fuel_type"].map(fuel_types)

In [66]:
autos["unrepaired_damage"].value_counts()

nein    33837
ja       4541
Name: unrepaired_damage, dtype: int64

Lastly, the `unrepaired_damage` column is very straightforward to translate.

In [67]:
autos["unrepaired_damage"] = autos["unrepaired_damage"].map({"ja": "yes", "nein": "no"})

To wrap up our cleaning, we look over the head of the cleaned data.

In [68]:
autos.head()

date_crawled                                               name  \
0 2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1 2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2 2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3 2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4 2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price   abtest   vehicle_type  registration_year    gearbox  power_ps  \
0  5000.0  control            van               2004     manual       158   
1  8500.0  control          sedan               1997  automatic       286   
2  8990.0     test          sedan               2009     manual       102   
3  4350.0  control     subcompact               2007  automatic        71   
4  1350.0     test  station wagon               2003     manual         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere     150000.0                   3       lpg     peugeot   
1     7er     150000.0                   6  gasoline         bmw   
2    golf      70000.0                   7  gasoline  volkswagen   
3  fortwo      70000.0                   6  gasoline       smart   
4   focus     150000.0                   7  gasoline        ford   

  unrepaired_damage ad_created  postal_code           last_seen  
0                no 2016-03-26        79588 2016-04-06 06:45:54  
1                no 2016-04-04        71034 2016-04-06 14:45:08  
2                no 2016-03-26        35394 2016-04-06 20:15:37  
3                no 2016-03-12        33729 2016-03-15 03:16:28  
4                no 2016-04-01        39218 2016-04-01 14:38:50

One potentially interesting further preprocessing step would be to analyze the `name` column for any keywords that could be extracted as new columns.

# Exploring Price and Mileage by Brand

Now that we have finished preprocessing the `autos` dataset, we can finally do some analysis. To start, we will do some basic aggregating by the `brand` column to explore differences in price and mileage between brands. Before performing any aggregations, we explore the `brand` column itself.

In [69]:
autos["brand"].nunique()

40

In [70]:
autos["brand"].value_counts(normalize = True)

volkswagen        0.211279
bmw               0.110052
opel              0.107567
mercedes_benz     0.096451
audi              0.086555
ford              0.069912
renault           0.047144
peugeot           0.029837
fiat              0.025639
seat              0.018271
skoda             0.016407
nissan            0.015272
mazda             0.015186
smart             0.014158
citroen           0.014008
toyota            0.012702
hyundai           0.010024
sonstige_autos    0.009853
volvo             0.009146
mini              0.008760
mitsubishi        0.008225
honda             0.007839
kia               0.007068
alfa_romeo        0.006640
porsche           0.006126
suzuki            0.005933
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002270
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001499
trabant           0.001392
r

In [71]:
autos["brand"].value_counts(normalize = True).head(10).sum()

0.8027073917792962

We see that while there are 40 different brands represented in this dataset, the top ten most common ones (Volkswagen, BMW, Opel, Mercedes Benz, Audi, Ford, Renault, Peugeot, Fiat, and Seat) make up almost 80% of the listings. The five brands are all German ones, and aside from Ford all of the top ten brands are European ones. This makes sense, since cars from German and other European brands would be the most readily available for German drivers to purchase and then potentially list on eBay when they are ready to sell their cars. For completeness, I will still aggregate over all 40 brands, but I will focus the majority of my analysis toward those in the top ten. To perform my aggregations, I will use the ["group by" features in Pandas](https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html). Additional info can once again be found in [VanderPlas's book](https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html).

In [75]:
top_10_brands = autos["brand"].value_counts().head(10).index

In [108]:
# Compute mean of odometer_km and price columns
# Round to the nearest whole number for easier interpretation
# Restrict focus to the top ten most common brands
# Sort from highest mean price to lowest
autos.groupby("brand")[["odometer_km", "price"]].mean().loc[top_10_brands].sort_values("price", ascending = False).round(0)

odometer_km   price
audi              129157.0  9337.0
mercedes_benz     130788.0  8628.0
bmw               132576.0  8571.0
volkswagen        128711.0  5604.0
seat              121131.0  4397.0
ford              124266.0  4055.0
peugeot           127154.0  3094.0
opel              129310.0  2975.0
fiat              117122.0  2814.0
renault           128071.0  2475.0

Among the top ten most common brands, there aren't any major differences between mean mileage, with all brands having a mean mileage between 115,000km and 135,000km. There are, however, some pretty striking differences in average price between brands. The three most expensive brands, Audi, Mercedes Benz, and BMW, are each about three times more expensive than the three least expensive brands, Renault, Fiat, and Opel. This difference in average price makes sense given the fact that the expensive brands are luxury ones which carry an associated price premium while the cheap brands make economy cars. The most popular brand, Volkswagen, sits in the middle ground between the two extremes, with an average listing price of \$5,604.

In [112]:
# Compute median of price column
# Round to the nearest whole number for easier interpretation
# Restrict focus to the top ten most common brands
# Sort from highest median price to lowest
autos.groupby("brand")["price"].agg(["min", "median", "max"]).loc[top_10_brands].sort_values("median", ascending = False).round(0)

min  median        max
audi           1.0  6350.0   175000.0
bmw            1.0  5900.0  1234566.0
mercedes_benz  1.0  5150.0   180000.0
volkswagen     1.0  2999.0   999999.0
seat           1.0  2750.0    33500.0
peugeot        1.0  2100.0    25500.0
ford           1.0  1800.0   999999.0
fiat           1.0  1650.0    25000.0
opel           1.0  1650.0    38990.0
renault        1.0  1350.0    93000.0

Since we noted that there are some unrealistically high prices in the data, even after filtering out the most egregiously unrealistic ones, we also can compare the median price between brands, since the median is less affected by outliers. The same pattern appears, though it is even more pronounced, with the median Audi price being almost quintuple the median Renault price.

# Exploring Price by Other Factors

In addition to comparing car prices across brands, we can also compare them across other factors. First, we will see if there are any patterns between price and mileage. Since we saw earlier that there is a fairly small number of distinct values in the `odometer_km` column, I will simply group by it directly.

In [88]:
autos.groupby("odometer_km")["price"].describe().sort_index().round(2)

count      mean        std   min     25%      50%       75%  \
odometer_km                                                                  
5000.0         786.0  13811.34  140954.99   1.0   444.0   1600.0   7500.00   
10000.0        241.0  20550.87   21041.49   1.0  7950.0  16500.0  27500.00   
20000.0        742.0  18448.48   23445.56  25.0  6800.0  14349.5  24200.00   
30000.0        760.0  16608.84   16869.42   1.0  7087.5  12700.0  21900.00   
40000.0        797.0  15499.57   13706.44  70.0  7000.0  12395.0  19995.00   
50000.0        994.0  15106.12   42744.97   1.0  5912.5  10500.0  17996.75   
60000.0       1128.0  12385.00   12952.66   1.0  4837.5   9225.0  16000.00   
70000.0       1187.0  10927.18    9329.35   1.0  4225.0   8250.0  15100.00   
80000.0       1375.0   9721.95    8877.64   1.0  3700.0   7499.0  13000.00   
90000.0       1673.0   8465.03    8543.22   1.0  2700.0   5995.0  11490.00   
100000.0      2058.0   8132.70   10680.58   1.0  2350.0   5600.0  10799.75   
125000.0      4858.0   6418.59   15929.97   1.0  1800.0   4000.0   8450.00   
150000.0     30088.0   3875.05   11789.11   1.0   999.0   2200.0   4800.00   

                   max  
odometer_km             
5000.0       3890000.0  
10000.0       163991.0  
20000.0       345000.0  
30000.0       220000.0  
40000.0       169999.0  
50000.0      1300000.0  
60000.0       190000.0  
70000.0        63999.0  
80000.0       105000.0  
90000.0        99900.0  
100000.0      299000.0  
125000.0      999999.0  
150000.0     1234566.0

In [117]:
autos.loc[autos["odometer_km"] < 50000, "price"].mean()

16377.964822609742

There is a very clear pattern when looking at both the mean and median price: as the odometer reading goes up, the price goes down. The dropoff is fairly minimal for cars with 5,000-50,000km; with the average prices for each odometer category hovering within a few thousand dollars of the overall average for that group (about \$16,000). After 50,000km, however, the average prices drop precipitously.

In [90]:
autos.fillna({"unrepaired_damage": "unknown"}).groupby("unrepaired_damage")["price"].describe().round(2)

count     mean       std  min     25%     50%     75%  \
unrepaired_damage                                                            
no                 33837.0  7346.33  25012.73  1.0  1800.0  4150.0  9000.0   
unknown             8309.0  3456.21  18264.49  1.0   650.0  1500.0  3700.0   
yes                 4541.0  2460.87  15228.91  1.0   500.0  1000.0  2500.0   

                         max  
unrepaired_damage             
no                 3890000.0  
unknown            1234566.0  
yes                 999990.0

Finally we compare prices between cars which have unrepaired damage and those which do not. I also replaced all of the null entries in the `unrepaired_damage` column with the value `"unknown"` so we can also consider prices for cars in that group. There is a very clear difference in average price between the three groups. Cars with unrepaired damage have an average price that is about a third of the average price for undamaged cars and cars with damage that has been repaired. In addition, even though cars with an unknown damage status have an average price about 50% higher than cars with unrepaired damage, that average price is still less than half the average price for cars without unrepaired damage.

# Exploring Popular Brand/Model Combinations

The last basic analysis we perform is identifying the most common brand/model combinations. This is very easy to do using the "group by" features in pandas.

In [134]:
# Group by brand and then get the value counts for the models in each brand
# Include null values in the value counts
# Then sort by values in descending order
autos.groupby("brand")["model"].value_counts(dropna = False).sort_values(ascending = False).head(10)

brand          model   
volkswagen     golf        3707
bmw            3er         2615
volkswagen     polo        1609
opel           corsa       1592
volkswagen     passat      1349
opel           astra       1348
audi           a4          1231
mercedes_benz  c_klasse    1136
bmw            5er         1132
mercedes_benz  e_klasse     958
Name: model, dtype: int64

Here we have the top ten most common brand/model combinations, which are all cars from German brands. If we wish to easily see what fraction of the overall dataset is represented by these common brand/model combinations, we [flatten the multi-indexed series](https://jakevdp.github.io/PythonDataScienceHandbook/03.05-hierarchical-indexing.html#Index-setting-and-resetting).

In [145]:
# Flatten multi-indexed series
flat_brand_model = autos.groupby("brand")["model"].value_counts(dropna = False).reset_index(name = "count")
flat_brand_model["proportion"] = flat_brand_model["count"]/autos.shape[0]
flat_brand_model.sort_values(["proportion"], ascending = False).head(10)

brand     model  count  proportion
297     volkswagen      golf   3707    0.079401
24             bmw       3er   2615    0.056011
298     volkswagen      polo   1609    0.034464
200           opel     corsa   1592    0.034099
299     volkswagen    passat   1349    0.028895
201           opel     astra   1348    0.028873
7             audi        a4   1231    0.026367
158  mercedes_benz  c_klasse   1136    0.024332
25             bmw       5er   1132    0.024247
159  mercedes_benz  e_klasse    958    0.020520

In [148]:
flat_brand_model["proportion"].sort_values(ascending = False).head(10).sum()

0.35720864480476366

Volkswagen Golfs and BMW 3 Series are by far the most common cars in the data set, and in total the top ten most common brand/model combinations make up about 36% of the listings in the set.

# Conclusion and Potential Next Steps

Over the course of this notebook, we went through the basic steps of preprocessing a dataset of used cars from eBay Kleinanzeigen. This involved loading the data into pandas; cleaning up the column names; and exploring the columns to identify ones which are likely candidates to drop immediately, ones which need to be converted into an appropriate type, and ones which contain potentially problematic values (null values and outliers). After processing the columns we identified, we then performed some basic analysis and aggregation to compare different categories of cars. In particular, we compared prices across different brands, odometer values, and states of repair. We also compared odometer readings across brands and explored the most common brand/model combinations.

Beyond the work we have done in this notebook, there are a number of potential next steps for interesting preprocessing and analysis. Some of them include the following.

- Analyze the `name` column for keywords that could be extracted and added as columns to the dataset.
- Categorize cars based on kilometers driven per year at time of listing and then analyze patterns in prices among those categories.
- Translate postal codes into geographic regions and compare the other variables (listing price, car brand/model/type) to see what differences or similarities there are between listings from various regions of Germany.
- Add data visualizations to further summarize the analysis.
- Explore various machine learning models to try and predict listing price based on the other variables.
- Try to identify interesting clusters in the data using unsupervised machine learning techniques.